### PR TITLE
perf: improved performance of `TabularDataset.__eq__` by a factor of up to 2

### PR DESCRIPTION
**Pull request by Marsmaennchen221**
### Summary of Changes

perf: improved performance of `TabularDataset.__eq__` by a factor of up to 2
perf: slightly improved performance of `TabularDataset.__hash__`
fix: corrected `TabularDataset.__sizeof__`

### CODE REVIEW

**Comment by Marsmaennchen221** [comment]
@lars-reimann do you really want to save the exact same data twice in a `TabularDataset`? Wouldn't it be better to not save the whole table and compact `_features`, `_target` and `_extras` if the user uses the `TabularDataset.to_table` method?

**Comment by lars-reimann** [comment]
Due to `pandas`'s copy-on-write semantics the current approach should barely need more memory. We only store some extra pointers.

We should add some performance benchmarks , though, to check runtime and memory use of the current implementation (not needed for this PR).

**Comment by lars-reimann** [comment]
Regarding column order: Should that matter for equality? If so, we should compare the tables and the names of target, features, and extras.

**Review by lars-reimann — CHANGES_REQUESTED** [review]

File: src/safeds/data/labeled/containers/_tabular_dataset.py [hunk 1]
```diff
@@ -116,7 +116,12 @@ def __sizeof__(self) -> int:
         size:
             Size of this object in bytes.
         """
-        return sys.getsizeof(self._target) + sys.getsizeof(self._features) + sys.getsizeof(self._table)
+        return (
+            sys.getsizeof(self._target)
+            + sys.getsizeof(self._features)
+            + sys.getsizeof(self._table)
+            + sys.getsizeof(self._extras)
```
[the remark this answers is not in the document]
Comment (Marsmaennchen221) [reply src/safeds/data/labeled/containers/_tabular_dataset.py:123]:
```suggestion
            sys.getsizeof(self._target)
            + sys.getsizeof(self._features)
            + sys.getsizeof(self._extras)
```

**Review by lars-reimann — APPROVED** [review]

**Pull request merged**